smartpower3/screens/settingscreen.cpp

```diff
@@ -412,70 +437,60 @@ void SettingScreen::changeSerialBaud(uint8_t baud_level)
-	}
+	changeLogInterval(log_interval, true);
 }
 
-void SettingScreen::selectSerialLogging(uint16_t color)
+void SettingScreen::selectBLLevel(uint16_t rectangle_color)
 {
-	for (int i = 1; i < SELECTION_BORDER_WIDTH+1; i++) {
-		tft->drawRect(x + X_LOG_LEVEL -i-50, y-1-i + Y_SERIAL -10, 135+i*2+120, 28+i*2+40, color);
+	for (int i = 0; i < SELECTION_BORDER_WIDTH; i++) {
+		tft->drawRect(x + X_BL_LEVEL-1-i, y+Y_BL_LEVEL-1-i, BL_LEVEL_WIDTH+2+i*2, BL_LEVEL_HEIGHT+2+i*2, rectangle_color);
 	}
 }
 
-void SettingScreen::selectLogInterval(uint16_t color)
+void SettingScreen::selectLogInterval(uint16_t text_color, uint16_t rectangle_color)
 {
-	com_log_interval->setTextColor(color, BG_COLOR);
-	com_log_interval->select();
+	com_log_interval->select(rectangle_color);
 }
 
-void SettingScreen::deSelectBLLevel(uint16_t color)
+void SettingScreen::deSelectBLLevel(uint16_t rectangle_color)
 {
-	for (int i = 1; i < SELECTION_BORDER_WIDTH+1; i++) {
-		tft->drawRect(x + X_BL_LEVEL -i, y-1-i, 135+i*2, 28+i*2, color);
+	for (int i = 1; i < SELECTION_BORDER_WIDTH; i++) {
+		tft->drawRect(x + X_BL_LEVEL-1-i, y+Y_BL_LEVEL-1-i, BL_LEVEL_WIDTH+2+i*2, BL_LEVEL_HEIGHT+2+i*2, rectangle_color);
 	}
 }
 
-void SettingScreen::deSelectLogInterval(uint16_t color)
+void SettingScreen::deSelectLogInterval(uint16_t text_color, uint16_t rectangle_color)
 {
-	com_log_interval->setTextColor(color, BG_COLOR);
-	drawLogIntervalValue(log_value[log_interval]);
-	com_log_interval->deSelect();
-}
-
-void SettingScreen::deSelectSerialLogging(uint16_t color)
-{
-	for (int i = 1; i < SELECTION_BORDER_WIDTH+1; i++) {
-		tft->drawRect(x + X_LOG_LEVEL -i-50, y-1-i + Y_SERIAL -10, 135+i*2+120, 28+i*2+40, color);
-	}
+	com_log_interval->deSelect(rectangle_color);
 }
 
 void SettingScreen::drawBacklightLevel(uint8_t level)
 {
-	tft->fillRect(x + X_BL_LEVEL + 2, y+1, 130, 24, BG_COLOR);
+	// clear inside space of level component
+	tft->fillRect(x + X_BL_LEVEL, y + Y_BL_LEVEL, BL_LEVEL_WIDTH, BL_LEVEL_HEIGHT, BG_COLOR);
 	for (int i = 0; i < level; i++) {
-		tft->fillRect(x + X_BL_LEVEL + 2 + (i*22), y + 1, 20, 24, TFT_YELLOW);
+		// draw individual level blocks
+		tft->fillRect(
+			// x start position + left inside space + (level_number * ((total_width - left_inside_space)/all_level_count))
+			x + X_BL_LEVEL + BL_LEVEL_SPACING + (i*((BL_LEVEL_WIDTH-BL_LEVEL_SPACING)/6)),
+			// y start_position + top_inside_space
+			y + Y_BL_LEVEL + BL_LEVEL_SPACING,
+			// block_width = ((total_width-left_inside_space)/all_level_count) - right_spacing_of_level_block
+			((BL_LEVEL_WIDTH-BL_LEVEL_SPACING)/6)-BL_LEVEL_SPACING,
+			// block_height = total_height - (top_spacing + bottom_spacing)
+			BL_LEVEL_HEIGHT-2*BL_LEVEL_SPACING,
+			TFT_GREEN);
 	}
 }
 
-void SettingScreen::selectSerialBaud(uint16_t color)
+void SettingScreen::selectSerialBaud(uint16_t text_color, uint16_t rectangle_color)
 {
-	com_serial_baud->setTextColor(color, BG_COLOR);
-	drawSerialBaud(this->serial_baud);
-	com_serial_baud->select();
```

Comment:
Unused color change of font when selected slows down screen redraw quite a lot - the selection movement then starts to behave erratically.